homeassistant/components/media_player/universal.py

```diff
@@ -0,0 +1,437 @@
+        # pylint: disable=too-many-arguments
+        self.hass = hass
+        self._name = name
+        self._children = children
+        self._cmds = commands
+        self._attrs = attributes
+        self._child_state = None
+
+        track_state_change(hass, self.dependencies, self.update_state)
+
+    def _entity_lkp(self, entity_id=None, state_attr=None):
+        """ Looks up an entity state from hass """
+        if entity_id is None:
+            return
+
+        state_obj = self.hass.states.get(entity_id)
+
+        if state_obj is None:
+            return
+
+        if state_attr:
+            return state_obj.attributes.get(state_attr)
+        return state_obj.state
+
+    def _override_or_child_attr(self, attr_name):
+        """ returns either the override or the active child for attr_name """
+        if attr_name in self._attrs:
+            return self._entity_lkp(*self._attrs[attr_name])
+
+        return self._child_attr(attr_name)
+
+    def _child_attr(self, attr_name):
+        """ returns the active child's attr """
+        active_child = self.active_child_state
+        return active_child.attributes.get(attr_name) if active_child else None
+
+    def _call_service(self, service_name, service_data=None,
+                      allow_override=False):
+        """ calls either a specified or active child's service """
+        if allow_override and service_name in self._cmds:
+            call_from_config(
+                self.hass, self._cmds[service_name], blocking=True)
+            return
+
+        if service_data is None:
+            service_data = {}
+
+        active_child = self.active_child_state
+        service_data[ATTR_ENTITY_ID] = active_child.entity_id
+
+        self.hass.services.call(DOMAIN, service_name, service_data,
+                                blocking=True)
+
+    @property
+    def should_poll(self):
+        """ Indicates whether HA should poll for updates """
+        return False
+
+    @property
+    def dependencies(self):
+        """ List of entity ids of entities that the mp depends on for state """
+        depend = copy(self._children)
+        for entity in self._attrs.values():
+            depend.append(entity[0])
+        return depend
+
+    @property
+    def master_state(self):
+        """ gets the master state from entity or none """
+        if CONF_STATE in self._attrs:
+            master_state = self._entity_lkp(*self._attrs[CONF_STATE])
+            return master_state if master_state else STATE_OFF
+        else:
+            return None
+
+    def _cache_active_child_state(self):
+        """ The state of the active child or None """
+        for child_name in self._children:
+            child_state = self.hass.states.get(child_name)
+            if child_state and child_state.state not in OFF_STATES:
+                self._child_state = child_state
+                return
+        self._child_state = None
+
+    @property
+    def active_child_state(self):
```

Comment:
If this is only used internally, no need to make a property as there is already an instance attribute.